pkg/util/subnet/subnetsupdater.go

```diff
@@ -0,0 +1,74 @@
+package subnet
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+import (
+	"strings"
+
+	mgmtnetwork "github.com/Azure/azure-sdk-for-go/services/network/mgmt/2020-08-01/network"
+	"github.com/Azure/go-autorest/autorest/to"
+)
+
+type Updater interface {
+	AddEndpointsToSubnets(endpoints []string, subnets []*mgmtnetwork.Subnet) (updatedSubnets []*mgmtnetwork.Subnet)
+}
+
+type UpdaterManager struct {
+}
+
+func (um *UpdaterManager) AddEndpointsToSubnets(endpoints []string, subnets []*mgmtnetwork.Subnet) (updatedSubnets []*mgmtnetwork.Subnet) {
+	for _, subnet := range subnets {
+		subnetChanged := addEndpointsToSubnet(endpoints, subnet)
+		if subnetChanged {
+			updatedSubnets = append(updatedSubnets, subnet)
+		}
+	}
+
+	return updatedSubnets
+}
+
+func addEndpointsToSubnet(endpoints []string, subnet *mgmtnetwork.Subnet) (subnetChanged bool) {
+	for _, endpoint := range endpoints {
+		endpointFound, serviceEndpointPtr := subnetContainsEndpoint(subnet, endpoint)
+		if !endpointFound {
+			addEndpointToSubnet(endpoint, subnet)
+			subnetChanged = true
+			continue
+		}
+
+		if serviceEndpointPtr.ProvisioningState != mgmtnetwork.Succeeded {
+			addEndpointToSubnet(endpoint, subnet)
+			subnetChanged = true
+		}
+	}
+
+	return subnetChanged
+}
+
+func subnetContainsEndpoint(subnet *mgmtnetwork.Subnet, endpoint string) (endpointFound bool, serviceEndpointPtr *mgmtnetwork.ServiceEndpointPropertiesFormat) {
+	if subnet == nil || subnet.ServiceEndpoints == nil {
+		return false, nil
+	}
+
+	for _, serviceEndpoint := range *subnet.ServiceEndpoints {
+		endpointFound = strings.EqualFold(*serviceEndpoint.Service, endpoint)
+		if endpointFound {
+			return true, &serviceEndpoint
+		}
+	}
+
+	return false, nil
+}
+
+func addEndpointToSubnet(endpoint string, subnet *mgmtnetwork.Subnet) {
+	if subnet.ServiceEndpoints == nil {
+		subnet.ServiceEndpoints = &[]mgmtnetwork.ServiceEndpointPropertiesFormat{}
+	}
+
+	serviceEndpoint := mgmtnetwork.ServiceEndpointPropertiesFormat{
+		Service:   to.StringPtr(endpoint),
+		Locations: &[]string{"*"},
+	}
+
+	*subnet.ServiceEndpoints = append(*subnet.ServiceEndpoints, serviceEndpoint)
+}
```

Comment:
Actually, you know what could be nice here ? having these functions be receiver functions on the subnet. 

for example :

```go
func (subnet mgmtnetwork.Subnet) containsEndpoint(endpoint string) (endpointFound bool, serviceEndpointPtr *mgmtnetwork.ServiceEndpointPropertiesFormat)
func (subnet *mgmtnetwork.Subnet) addEndpoint(endpoint string)
```
then you can even remove the pointer check on subnet for `containsEndpoint` because we are not using a pointer anymore

Reply:
I replied too fast.

I agree it was a good idea but we can't do that as  "mgmtnetwork.Subnet" is not a struct we define, it is on "vendor/github.com/Azure/azure-sdk-for-go/services/network/mgmt/2020-08-01/network/models.go" so we can't do it.

Reply:
I thought it was a good idea but I just realized that "mgmtnetwork.Subnet" is from "vendor/github.com/Azure/azure-sdk-for-go/services/network/mgmt/2020-08-01/network/models.go" so I think it is not that easy to change in this case. 

If the struct was defined in this module (ARO-RP) the change would be easy and straightforward but knowing that is from another module I would not touch it for now. 

Right? 🙂

Reply:
yeah this is what I said in the edit afterwards. there isn't a good way to do it except wrapping it into an other struct. which is probably not worth it.

Reply:
Either I answered previously to your edit or I missed it. But know we agree on that 🙂